### PR TITLE
Web Inspector: URL breakpoints should also pause when HTML attributes that trigger loads are set

### PR DESCRIPTION
**Pull request by dcrousso**
#### 7ad0216f7c41bea64fd4d315e9df4d020c9f0830
<pre>
Web Inspector: URL breakpoints should also pause when HTML attributes that trigger loads are set
<a href="https://bugs.webkit.org/show_bug.cgi?id=218140">https://bugs.webkit.org/show_bug.cgi?id=218140</a>
&lt;rdar://problem/70636814&gt;

Reviewed by Patrick Angle.

This will give developers even more power over understanding the cause of network requests.

It&apos;s still the case that by the time the URL breakpoint pauses, we&apos;ve already configured the network
request (i.e. it cannot be changed), but at least this way a developer can pause when the request is
made, rather than inferring that by looking at the backtrace afterwards.

* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::willSendRequestImpl):
(WebCore::InspectorInstrumentation::willSendRequestOfTypeImpl):
* Source/WebCore/inspector/agents/InspectorDOMDebuggerAgent.h:
* Source/WebCore/inspector/agents/InspectorDOMDebuggerAgent.cpp:
(WebCore::InspectorDOMDebuggerAgent::willSendRequest): Added.
(WebCore::InspectorDOMDebuggerAgent::willSendRequestOfType): Added.
(WebCore::InspectorDOMDebuggerAgent::breakOnURLIfNeeded):
(WebCore::InspectorDOMDebuggerAgent::willSendXMLHttpRequest):
(WebCore::InspectorDOMDebuggerAgent::willFetch):
Leverage existing instrumentation to notify the `DOMDebugger` whenever a new network request is made,
at which point the it can check for any existing URL breakpoints and pause if one is found.

* Source/JavaScriptCore/inspector/protocol/Debugger.json:
* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.pauseReasonFromPayload):
* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
(WI.SourcesNavigationSidebarPanel.prototype._updatePauseReasonSection):
Merge the `&quot;XHR&quot;` and `&quot;Fetch&quot;` pause reasons into a single `&quot;URL&quot;` since URL breakpoints can now be
triggered by more than just those two APIs (i.e. DOM APIs too).

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::shouldContinueAfterNotifyingLoadedFromMemoryCache):
Make sure to also copy over the `ResourceRequest::Requester`.

* Source/WebCore/dom/ScriptDisallowedScope.h:
(WebCore::ScriptDisallowedScope::isScriptAllowedInMainThread): Renamed from `isEventAllowedInMainThread`.
* Source/WebCore/dom/EventTarget.cpp:
(WebCore::EventTarget::fireEventListeners):
Drive-by: Rename to make it less specific to events, as the same logic can be useful elsewhere.

* LayoutTests/inspector/dom-debugger/resources/dataDOM.json: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-dom-all-requests.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-dom-all-requests-expected.txt: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-dom-containing.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-dom-containing-expected.txt: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-dom-matching.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-dom-matching-expected.txt: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-dom.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-dom-expected.txt: Added.

* LayoutTests/inspector/dom-debugger/url-breakpoints-fetch-all-requests.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-fetch-all-requests-expected.txt: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-fetch-containing.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-fetch-containing-expected.txt: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-fetch-matching.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-fetch-matching-expected.txt: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-fetch.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-fetch-expected.txt: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-xhr-all-requests.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-xhr-all-requests-expected.txt: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-xhr-containing.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-xhr-containing-expected.txt: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-xhr-matching.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-xhr-matching-expected.txt: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-xhr.html: Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-xhr-expected.txt: Added.
* LayoutTests/inspector/worker/dom-debugger-url-breakpoints-fetch.html: Added.
* LayoutTests/inspector/worker/dom-debugger-url-breakpoints-fetch-expected.txt: Added.
* LayoutTests/inspector/worker/dom-debugger-url-breakpoints-xhr.html: Added.
* LayoutTests/inspector/worker/dom-debugger-url-breakpoints-xhr-expected.txt: Added.
* LayoutTests/inspector/worker/resources/dataDOM.json: Added.
* LayoutTests/inspector/worker/resources/resource-utilities.js:
(loadResourceDOM): Added.
* LayoutTests/inspector/worker/resources/worker-dom-debugger.js:
(const.functions.triggerXHRRequest): Renamed from `triggerXHR`.
(const.functions.triggerFetchRequest): Renamed from `triggerFetch`.
(const.functions.triggerDOMRequest): Added.
* LayoutTests/inspector/dom-debugger/url-breakpoints-all-requests.html: Removed.
* LayoutTests/inspector/dom-debugger/url-breakpoints-all-requests-expected.txt: Removed.
* LayoutTests/inspector/dom-debugger/url-breakpoints-containing.html: Removed.
* LayoutTests/inspector/dom-debugger/url-breakpoints-containing-expected.txt: Removed.
* LayoutTests/inspector/dom-debugger/url-breakpoints-matching.html: Removed.
* LayoutTests/inspector/dom-debugger/url-breakpoints-matching-expected.txt: Removed.
* LayoutTests/inspector/dom-debugger/url-breakpoints.html: Removed.
* LayoutTests/inspector/dom-debugger/url-breakpoints-expected.txt: Removed.
* LayoutTests/inspector/worker/dom-debugger-url-breakpoints.html: Removed.
* LayoutTests/inspector/worker/dom-debugger-url-breakpoints-expected.txt: Removed.
Split URL breakpoint tests into separate files for each API (i.e. one for XHR, one for `fetch`, etc.).

Canonical link: <a href="https://commits.webkit.org/254488@main">https://commits.webkit.org/254488@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/3bc6c6747d7564736f3a2af54e9917514667a6f1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89192 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33753 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20012 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98511 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154824 "Hash 3bc6c674 for PR 4239 does not build (failure)") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32261 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27809 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81554 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92980 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25626 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76117 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25559 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80491 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/80466 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68533 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30040 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14528 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/74693 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29764 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15492 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26325 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3150 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33211 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38452 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/77558 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31912 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34583 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17161 "Passed tests") |

### CODE REVIEW

File: Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js [hunk 1]
```diff
@@ -315,7 +315,6 @@ WI.DebuggerManager = class DebuggerManager extends WI.Object
             return WI.DebuggerManager.PauseReason.EventListener;
         case InspectorBackend.Enum.Debugger.PausedReason.Exception:
             return WI.DebuggerManager.PauseReason.Exception;
-        case InspectorBackend.Enum.Debugger.PausedReason.Fetch:
             return WI.DebuggerManager.PauseReason.Fetch;
```
Comment (patrickangle) [inline Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js]: Did the return statement need to be removed as well? 😅

File: Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js [hunk 2]
```diff
@@ -331,8 +330,10 @@ WI.DebuggerManager = class DebuggerManager extends WI.Object
             return WI.DebuggerManager.PauseReason.Timeout;
         case InspectorBackend.Enum.Debugger.PausedReason.Timer:
             return WI.DebuggerManager.PauseReason.Timer;
-        case InspectorBackend.Enum.Debugger.PausedReason.XHR:
-            return WI.DebuggerManager.PauseReason.XHR;
+        case InspectorBackend.Enum.Debugger.PausedReason.URL:
+        case InspectorBackend.Enum.Debugger.PausedReason.Fetch: // COMPATIBILITY (macOS 13.0, iOS 16.0): Debugger.paused.reason.Fetch was replaced by Debugger.paused.reason.URL
+        case InspectorBackend.Enum.Debugger.PausedReason.XHR: // COMPATIBILITY (macOS 13.0, iOS 16.0): Debugger.paused.reason.XHR was replaced by Debugger.paused.reason.URL
```
Comment (patrickangle) [inline Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js]: Unless I'm mistaken these will effectively be `case undefined:` when the new protocol doesn't include them. While probably not a problem, it is a bit strange to think about.

**Review by patrickangle — APPROVED** [review]

File: Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js (hunk 1 above)
Comment (dcrousso) [reply Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js]: oh lol oops yes will remove

File: Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js (hunk 2 above)
Comment (dcrousso) [reply Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js]: yeah it's maybe a bit odd, but this is definitely not the first time we've done this 😅

**Pull request closed without merging**